axon/threads.go

```diff
@@ -6,247 +6,183 @@ package axon
 
 import (
 	"fmt"
+	"log"
+	"math"
 	"runtime"
 	"sort"
+	"sync"
 
 	"github.com/emer/emergent/timer"
 	"github.com/goki/ki/ints"
 )
 
-// NetThread specifies how to allocate threads & chunks to each task, and
-// manages running those threads (goroutines)
-type NetThread struct {
-	NThreads  int     `desc:"number of parallel threads to deploy"`
-	ChunksPer int     `desc:"number of chunks per thread to use -- each thread greedily grabs chunks"`
-	Work      WorkMgr `view:"-" desc:"work manager"`
-}
-
-func (th *NetThread) Set(nthr, chk int) {
-	maxP := runtime.GOMAXPROCS(0)
-	th.NThreads = nthr
-	if th.NThreads < 1 {
-		th.NThreads = 1
+// Maps the given function across the [0, total) range of items, using
+// nThreads goroutines.
+func parallelRun(fun func(st, ed int), total int, nThreads int) {
+	itemsPerThr := int(math.Ceil(float64(total) / float64(nThreads)))
+	waitGroup := sync.WaitGroup{}
+	for start := 0; start < total; start += itemsPerThr {
+		start := start // be extra sure with closure
+		end := ints.MinInt(start+itemsPerThr, total)
+		waitGroup.Add(1) // todo: move out of loop
+		go func(st, ed int) {
+			fun(st, ed)
+			waitGroup.Done()
+		}(start, end)
 	}
-	if th.NThreads > maxP {
-		th.NThreads = maxP
-	}
-	th.ChunksPer = chk
+	waitGroup.Wait()
 }
 
-func (th *NetThread) Alloc(tot int) {
-	th.Work.Alloc(tot, th.NThreads, th.ChunksPer)
+// NetThreads parameterizes how many goroutines to use for each task
+type NetThreads struct {
+	Neurons   int `desc:"for basic neuron-level computation -- highly parallel and linear in memory -- should be able to use a lot of threads"`
+	SendSpike int `desc:"for sending spikes per neuron -- very large memory footprint through synapses and is very sparse -- suffers from significant bandwidth limitations -- use fewer threads"`
+	SynCa     int `desc:"for synaptic-level calcium updating -- very large memory footprint through synapses but linear in order -- use medium number of threads"`
 }
 
-func (th *NetThread) Run(fun func(st, ed int)) {
-	th.Work.Run(th.NThreads, fun)
-}
+// SetDefaults uses heuristics to determine the number of goroutines to use
+// for each task: Neurons, SendSpike, SynCa.
+func (nt *NetThreads) SetDefaults(nNeurons, nPrjns, nLayers int) {
+	maxProcs := runtime.GOMAXPROCS(0) // query GOMAXPROCS
```

Comment:
Do you mean the `query GOMAXPROCS`? Because I remember both of us had to look up what that actually does during our last pair coding, so I thought I'd add a comment.

Reply:
Hmm. Maybe it's only obvious to me now after we looked it up.
But I feel like assigning the result to a variable named `maxProcs` makes it clear what it's doing without the comment.